bin/propolis-server/src/lib/stats.rs

```diff
@@ -135,57 +164,77 @@ pub async fn start_oximeter_server(
         interval: OXIMETER_STAT_INTERVAL,
     };
 
+    // Create a child logger, to avoid intermingling the producer server output
+    // with the main Propolis server.
+    let producer_log = oximeter_producer::LogConfig::Logger(
+        log.new(slog::o!("component" => "oximeter-producer")),
+    );
     let config = Config {
         server_info,
         registration_address,
         dropshot: dropshot_config,
-        log: oximeter_producer::LogConfig::Config(logging_config),
+        log: producer_log,
     };
 
-    const N_ATTEMPTS: u8 = 2;
-    const RETRY_WAIT_SEC: u64 = 1;
-    for _ in 0..N_ATTEMPTS {
-        let server = Server::start(&config).await;
-        match server {
-            Ok(server) => {
-                info!(
-                    log,
-                    "connected {:?} to oximeter {:?}",
-                    my_address,
-                    registration_address
-                );
-                return Some(server);
-            }
-            Err(e) => {
-                warn!(
-                    log,
-                    "Could not connect to oximeter (retrying in {}s):\n{}",
-                    RETRY_WAIT_SEC,
-                    e,
-                );
-
-                tokio::time::sleep(tokio::time::Duration::from_secs(
-                    RETRY_WAIT_SEC,
-                ))
-                .await;
-            }
-        }
-    }
-    error!(log, "Could not connect to oximeter after {} attempts", N_ATTEMPTS);
-
-    None
+    // Create the server which will attempt to register with Nexus.
+    Server::with_registry(registry.clone(), &config).await
 }
 
 /// Creates and registers a set of server-level metrics for an instance.
-pub fn register_server_metrics(
-    id: Uuid,
-    server: &Server,
+///
+/// This attempts to initialize kstat-based metrics for vCPU usage data. This
+/// may fail, in which case those metrics will be unavailable.
+pub async fn register_server_metrics(
+    registry: &ProducerRegistry,
+    virtual_machine: VirtualMachine,
+    #[cfg(all(not(test), target_os = "illumos"))] log: &Logger,
+    #[cfg(not(all(not(test), target_os = "illumos")))] _log: &Logger,
```

Comment:
nit: Please let's not have `cfg()` soup inside the function arguments

Reply:
Yeah, this was my annoyance showing through. I'll host it up somewhere.

Reply:
Move to the function itself in d61fd69